processing/src/main/java/io/druid/query/groupby/epinephelinae/ByteBufferHashTable.java

```diff
@@ -0,0 +1,380 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.groupby.epinephelinae;
+
+import com.google.common.primitives.Ints;
+import io.druid.java.util.common.IAE;
+import io.druid.java.util.common.ISE;
+
+import java.nio.ByteBuffer;
+
+public class ByteBufferHashTable
+{
+  public static int calculateTableArenaSizeWithPerBucketAdditionalSize(
+      int bufferCapacity,
+      int bucketSize,
+      int perBucketAdditionalSize
+  )
+  {
+    return (bufferCapacity / (bucketSize + perBucketAdditionalSize)) * bucketSize;
+  }
+
+  public static int calculateTableArenaSizeWithFixedAdditionalSize(
+      int bufferCapacity,
+      int bucketSize,
+      int fixedAdditionalSize
+  )
+  {
+    return ((bufferCapacity - fixedAdditionalSize) / bucketSize) * bucketSize;
+  }
+
+  protected final int maxSizeForTesting; // Integer.MAX_VALUE in production, only used for unit tests
+
+  protected static final int HASH_SIZE = Ints.BYTES;
+
+  protected final float maxLoadFactor;
+  protected final int initialBuckets;
+  protected ByteBuffer buffer;
+  protected int bucketSizeWithHash;
+  protected int tableArenaSize;
+  protected int keySize;
+
+  protected int tableStart;
+
+  // Buffer pointing to the current table (it moves around as the table grows)
+  protected ByteBuffer tableBuffer;
+
+  // Number of elements in the table right now
+  protected int size;
+
+  // Maximum number of elements in the table before it must be resized
+  protected int maxSize;
+
+  // current number of available/used buckets in the table
+  protected int buckets;
```

Comment:
may be also document in the comment for above two variables that these numbers change on table resize.